qiskit/algorithms/minimum_eigensolvers/vqe.py

```diff
@@ -0,0 +1,343 @@
+
+            values = estimator_result.values
+
+            if self.callback is not None:
+                metadata = estimator_result.metadata
+                for params, value, meta in zip(parameters, values, metadata):
+                    eval_count += 1
+                    self.callback(eval_count, params, value, meta)
+
+            energy = values[0] if len(values) == 1 else values
+
+            return energy
+
+        return evaluate_energy
+
+    def _get_evaluate_gradient(
+        self,
+        ansatz: QuantumCircuit,
+        operator: BaseOperator | PauliSumOp,
+    ) -> tuple[Callable[[np.ndarray], np.ndarray]]:
+        """Get a function handle to evaluate the gradient at given parameters for the ansatz.
+
+        Args:
+            ansatz: The ansatz preparing the quantum state.
+            operator: The operator whose energy to evaluate.
+
+        Returns:
+            A function handle to evaluate the gradient at given parameters for the ansatz.
+
+        Raises:
+            AlgorithmError: If the primitive job to evaluate the gradient fails.
+        """
+
+        def evaluate_gradient(parameters: np.ndarray) -> np.ndarray:
+            # broadcasting not required for the estimator gradients
+            try:
+                job = self.gradient.run([ansatz], [operator], [parameters])
+                gradients = job.result().gradients
+            except Exception as exc:
+                raise AlgorithmError("The primitive job to evaluate the gradient failed!") from exc
+
+            return gradients[0]
+
+        return evaluate_gradient
+
+    def _check_operator_ansatz(self, operator: BaseOperator | PauliSumOp) -> QuantumCircuit:
+        """Check that the number of qubits of operator and ansatz match and that the ansatz is
+        parameterized.
+        """
+        # avoid changing the original ansatz as a user may not expect this behavior
+        ansatz = self.ansatz.copy()
+
+        if operator.num_qubits != ansatz.num_qubits:
+            try:
+                logger.info(
+                    "Trying to resize ansatz to match operator on %s qubits.", operator.num_qubits
+                )
+                ansatz.num_qubits = operator.num_qubits
+            except AttributeError as error:
+                raise AlgorithmError(
+                    "The number of qubits of the ansatz does not match the "
+                    "operator, and the ansatz does not allow setting the "
+                    "number of qubits using `num_qubits`."
+                ) from error
+
+        if ansatz.num_parameters == 0:
+            raise AlgorithmError("The ansatz must be parameterized, but has no free parameters.")
+
+        return ansatz
+
+
+def _build_vqe_result(
+    optimizer_result: OptimizerResult,
+    eval_time: float,
+    ansatz: QuantumCircuit,
+    aux_values: ListOrDict[tuple[complex, tuple[complex, int]]],
+) -> VQEResult:
+    result = VQEResult()
+    result.eigenvalue = optimizer_result.fun
```

Comment:
What did we conclude over the main eigenvalue and returning variance/metadata - like my comment on OptimzerResult we could put the whole Estimator Result in here and let people get at whatever is in it.

Reply:
We looked into this possibility and, to our understanding, `evaluate_energy()` (the function passed over to the estimator) is expected by the optimizer to only return the energy value, not the metadata.
In order to return the estimator metadata together with the evaluated value we would have to resort to some kind of history-saving mechanism similar to how the current callback is called. At this point, I would say, having an additional `history` that can already be retrieved through the callback is a bit redundant. What we thought would make sense is to modify the callback so that it returns the whole metadata and not `std/variance/shots`.